### ISSUE TITLE
pc.download_files()

### ISSUE THREAD

**Issue by jq6l43d1** [reflex-dev/reflex#909]
We have pc.upload_files() for uploading files. I think we should also have pc.download_files() for downloading files. I don't know of a clean way to create a download link in pynecone. We should have one. #795

**Comment by Alek99** [comment]
Nice suggestion we will start looking in  on how to implement this feature!

**Comment by Justin900429** [comment]
@jq6l43d1 

An interesting finding is to use api route provided [here](https://pynecone.io/docs/advanced-guide/api-routes).


```python
import starlette
...
async def file_download(filename: str):
    return "<to your file path>"
...
app = pc.State(State)
app.api.add_api_route("/download/{filename}", file_download, response_class=starlette.responses.FileResponse)
app.compile()
...
```

By doing this, you can assess `localhost:8000` to get the file. You can use `pc.redirect` to access the download link. However, this method can be tedious and impractical due to different ports used (3000 for main program and 8000 for api server).

Hope this can be helpful for your current issues :)

**Comment by Justin900429** [comment]
Actually these two are the same rendered by FastAPI in this scenario. But the method you provided can be more flexible to different file types or even return buffer. Below is my current use case.

```python
@app.api.get("/download/{filename}")
async def file_download(filename: str, background_tasks: BackgroundTasks):
    buffer = BytesIO()

    local_filename = "zipfiles/" + filename + ".zip"

    # Open a zipfile and write into to the buffer
    with open(local_filename, "rb") as zip_file:
        buffer.write(zip_file.read())
    buffer.seek(0)

    background_tasks.add_task(buffer.close)
    background_tasks.add_task(os.remove, path=local_filename)
    headers = {
        "Content-Disposition": f'attachment; filename="{filename}.zip"'
    }
    return Response(buffer.getvalue(), headers=headers, media_type="application/zip")
```

Thanks for sharing your method!

**Comment by TaiJuWu** [comment]
Does this feature need frontend?
I have implemented backend on https://github.com/TaiJuWu/pynecone/tree/download
Example code:
```python
app = pc.App(state=State)
app.add_page(index)
app.add_download_file("/test1", "/Users/tai-juwu/ws/pynecone/download/test.JPG")
app.add_download_file("/test2", "/Users/tai-juwu/ws/pynecone/download/test.pdf")
app.add_download_file("/test3", "/Users/tai-juwu/ws/pynecone/download/test.mp4")
app.compile()
```

I meet some problem is I try to call app.add_download page in app.compile but it failed.

**Comment by Justin900429** [comment]
@TaiJuWu  No frontend is needed. You can left index page blank and just implment the backend service.
The ony difference here is that the ip address should be accessed with port **8000**. You can modify the port within `pconfig.py`.

Back to your example code, the default `pc.App` has no method called `add_download_file`. If you would like to add a new route, you can try `app.api.add_route(...)`. One advantage of using `api.add_route` is that you don't have to specify the route one by one like you did in your example code. You can directly use a dump name and send into the function like below:

```python
import starlette  # Already installed with Pynecone!
...
async def file_download(filename: str):
    return "<your true file path>"
...
app = pc.State(State)
app.api.add_api_route("/download/{filename}", file_download, response_class=starlette.responses.FileResponse)
app.compile()
...
```

Hope this helps 💪

**Comment by Lendemor** [comment]
Solved by #1797
